hudi-spark-datasource/hudi-spark-common/src/main/scala/org/apache/hudi/RecordLevelIndexSupport.scala

```diff
@@ -122,47 +123,78 @@ object RecordLevelIndexSupport {
   val INDEX_NAME = "RECORD_LEVEL"
 
   /**
-   * If the input query is an EqualTo or IN query on simple record key columns, the function returns a tuple of
+   * If the input query is an EqualTo or IN query on record key columns, the function returns a tuple of
    * list of the query and list of record key literals present in the query otherwise returns an empty option.
    *
    * @param queryFilter The query that need to be filtered.
    * @return Tuple of filtered query and list of record key literals that need to be matched
    */
-  def filterQueryWithRecordKey(queryFilter: Expression, recordKeyOpt: Option[String]): Option[(Expression, List[String])] = {
+  def filterQueryWithRecordKey(queryFilter: Expression, recordKeyFields: Array[String]): Option[(Expression, List[String])] = {
+    val isComplexRecordKey = recordKeyFields.length > 1
     queryFilter match {
+      // Handle EqualTo expressions
       case equalToQuery: EqualTo =>
         val attributeLiteralTuple = getAttributeLiteralTuple(equalToQuery.left, equalToQuery.right).orNull
         if (attributeLiteralTuple != null) {
           val attribute = attributeLiteralTuple._1
           val literal = attributeLiteralTuple._2
-          if (attribute != null && attribute.name != null && attributeMatchesRecordKey(attribute.name, recordKeyOpt)) {
-            Option.apply(equalToQuery, List.apply(literal.value.toString))
+          if (attribute != null && attribute.name != null && attributeMatchesRecordKey(attribute.name, recordKeyFields)) {
+            val recordKeyLiteral = getRecordKeyLiteral(attribute, literal, isComplexRecordKey)
+            Option.apply(equalToQuery, List.apply(recordKeyLiteral))
           } else {
             Option.empty
           }
         } else {
           Option.empty
         }
 
+      // Handle In expressions
       case inQuery: In =>
         var validINQuery = true
         inQuery.value match {
           case attribute: AttributeReference =>
-            if (!attributeMatchesRecordKey(attribute.name, recordKeyOpt)) {
+            if (!attributeMatchesRecordKey(attribute.name, recordKeyFields)) {
               validINQuery = false
             }
           case _ => validINQuery = false
         }
         var literals: List[String] = List.empty
         inQuery.list.foreach {
-          case literal: Literal => literals = literals :+ literal.value.toString
+          case literal: Literal => literals = literals :+ getRecordKeyLiteral(inQuery.value.asInstanceOf[AttributeReference], literal, isComplexRecordKey)
```

Comment:
We need to check all the record key fields are got involved.

Reply:
Addressed